### PR TITLE
boards/opentitan: tests: otbn: Remove global static buffers

### PR DESCRIPTION
**Pull request by thulithwilfred**
Signed-off-by: Wilfred Mallawa <wilfred.mallawa@wdc.com>

### Pull Request Overview
Removed global static buffers from the OTBN test.

### Testing Strategy
`make test`

`make ci-all`

### Documentation Updated
- [x] No updates are required.

### Formatting
- [x] Ran `make fmt`.

### CODE REVIEW

File: boards/opentitan/src/tests/otbn.rs [hunk 1]
```diff
@@ -81,10 +82,26 @@ extern "C" {
     static _eapps: u8;
 }
 
+/// Static init an OtbnTestCallback, with
+/// respective buffers allocated
+/// Note: static_init!() returns an &'static mut reference.
+unsafe fn static_init_test_cb() -> &'static OtbnTestCallback {
+    let output_buf = static_init!([u8; 1024], [0; 1024]);
+    let temp_buf = static_init!([u8; 4], [0; 4]);
```
Comment (alistair23) [inline boards/opentitan/src/tests/otbn.rs:90]: You have identified a bug in the current code here.

The `TEMP` buffer is only ever used with `load_data()`. Previously `load_data()` took a `'static mut` as it was an async (like most calls in Tock). It was recently converted to just be a synchronous call as it's only used for the OTBN and it loads the data synchronously.

This means that `TEMP` and the new `temp_buf` don't need to be static. So you can remove this and just create a local array in `otbn_run_rsa_binary()` with

```rust
let temp_buf = [0; 4];
```

and just use that directly

File: boards/opentitan/src/tests/otbn.rs [hunk 2]
```diff
@@ -81,10 +82,26 @@ extern "C" {
...
+        OtbnTestCallback::new(output_buf, temp_buf)
+    )
+}
+
 #[test_case]
 fn otbn_run_rsa_binary() {
     let perf = unsafe { PERIPHERALS.unwrap() };
     let otbn = &perf.otbn;
+    let cb = unsafe { static_init_test_cb() };
+    let output = cb.output_buf.take().unwrap();
```
Comment (alistair23) [inline boards/opentitan/src/tests/otbn.rs:103]: You need to save this again in `op_done()`

**Review by alistair23 — CHANGES_REQUESTED** [review]

File: boards/opentitan/src/tests/otbn.rs (hunk 2 above)
Comment (thulithwilfred) [reply boards/opentitan/src/tests/otbn.rs:99]: Thanks, I have made these changes.

File: boards/opentitan/src/tests/otbn.rs [hunk 3]
```diff
@@ -66,13 +67,13 @@ impl<'a> OtbnTestCallback {
 impl<'a> Client<'a> for OtbnTestCallback {
     fn op_done(&'a self, result: Result<(), ErrorCode>, output: &'static mut [u8]) {
         self.op_done.set(true);
+        self.output_buf.replace(output);
         assert_eq!(result, Ok(()));
-        assert_eq!(output[0..256], EXPECTING);
+        // Use this copy of output for assertion
+        assert_eq!(self.output_buf.take().unwrap()[0..256], EXPECTING);
```
Comment (alistair23) [inline boards/opentitan/src/tests/otbn.rs:73]: Close, but this won't work.

You correctly replace the `self.output_buf` above, but then you take it here and it gets dropped.

You would instead want to do:

```rust
assert_eq!(output[0..256], EXPECTING);
self.output_buf.replace(output);
```

File: boards/opentitan/src/tests/otbn.rs [hunk 4]
```diff
@@ -101,32 +114,30 @@ fn otbn_run_rsa_binary() {
...
         run_kernel_op(1000);
 
         // Set the RSA mode to encryption
         // The address is the offset of `mode` in the RSA elf
-        unsafe {
-            TEMP[0] = 1;
-            assert_eq!(otbn.load_data(0, &TEMP[0..4]), Ok(()));
-        }
+        temp[0] = 1;
+        assert_eq!(otbn.load_data(0, &temp[0..4]), Ok(()));
```
Comment (alistair23) [inline boards/opentitan/src/tests/otbn.rs:132]: You should be able to replace the `&temp[0..4]` with just `temp`

**Review by alistair23 — CHANGES_REQUESTED** [review]

**Review by alistair23 — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

**Pull request merged**